ch5/ex-5-17-qsort-fields.c

```diff
@@ -0,0 +1,228 @@
+
+	parseargs(argc, argv);
+
+	if ((nlines = readlines(&buff, &lines)) == LNS_ERROR) {
+		printf("input too big to sort\n");
+		return 0;
+	}
+
+	linedata = malloc(nlines * sizeof(char **));
+	for (i = 0; i < nlines; i++)
+		linedata[i] = getsortdata(lines[i]);
+
+	quicksort((void **)linedata, 0, nlines - 1, (VOIDCOMP)compare);
+
+	for (i = 0; i < nlines; i++)
+		lines[i] = line(linedata[i]);
+
+	writelines(lines, nlines);
+
+	freestuff(linedata, nlines);
+	freelines(buff, lines);
+	return 0;
+}
+
+int compare(char **sortdata1, char **sortdata2)
+{
+	int i, rslt = 0;
+
+	for (i = 0; i < nsortflds && rslt == 0; i++)
+		rslt = (*compares[i])(sortdata1[i], sortdata2[i]);
+	return rslt;
+}
+
+void parseargs(int argc, char *argv[])
+{
+	int i;
+	int maxsortcount = argc - 1;
+
+	nsortflds = 0;
+	sortidxs = malloc(maxsortcount * sizeof(int));
+	compares = malloc(maxsortcount * sizeof(CHARCOMP));
+	folds = malloc(maxsortcount * sizeof(int));
+	dirsorts = malloc(maxsortcount * sizeof(int));
+
+	argv++;
+	for (i = 0; i < argc - 1; i++) {
+		char *arg = argv[i];
+
+		int fieldidx;
+		int numeric = 0;
+		int reverse = 0;
+		int (*compare)(char *, char *);
+		int fold = 0;
+		int dirsort = 0;
+
+		if (*arg == '-' && isdigit(*++arg)) {
+			fieldidx = atoi(arg) - 1;
+			while (*++arg) {
+				if (*arg == 'n')
+					numeric = 1;
+				else if (*arg == 'r')
+					reverse = 1;
+				else if (*arg == 'f')
+					fold = 1;
+				else if (*arg == 'd')
+					dirsort = 1;
+			}
+			if (numeric && reverse)
+				compare = numcmpr;
+			else if (numeric)
+				compare = numcmp;
+			else if (reverse)
+				compare = strcmpr;
+			else
+				compare = (CHARCOMP)strcmp;
+			maxsortidx =
+				fieldidx > maxsortidx ? fieldidx : maxsortidx;
+			sortidxs[nsortflds] = fieldidx;
+			compares[nsortflds] = compare;
+			folds[nsortflds] = fold;
+			dirsorts[nsortflds] = dirsort;
+			nsortflds++;
+		}
+	}
+}
+
+char **getsortdata(char *line)
+{
+	char *chr, *field;
+	int fieldidx, i;
+
+	char *fieldspace = malloc((strlen(line) + 1) * sizeof(char));
+	char **tmpfields = malloc((maxsortidx + 2) * sizeof(char *));
+	char **sortdata = malloc((nsortflds + 2) * sizeof(char *));
+
+	strcpy(fieldspace, line);
+
+	/* split line into fields and create temporary pointer for each field */
+	field = fieldspace;
+	fieldidx = 0;
+	for (chr = fieldspace; *chr; chr++) {
+		if (*chr == '\t' || *chr == '\n') {
+			*chr = '\0';
+			/*maxsortidx + 1 so last sort field is terminated */
+			if (fieldidx <= maxsortidx + 1)
+				tmpfields[fieldidx] = field;
+			field = chr + 1;
+			fieldidx++;
+		}
+	}
+
+	/* ptr to sort-by fields in priority order, apply -f and -d if needed */
+	for (i = 0; i < nsortflds; i++) {
+		field = tmpfields[sortidxs[i]];
+		if (folds[i])
+			fold(field);
+		if (dirsorts[i])
+			dirsort(field);
+		sortdata[i] = field;
+	}
+	sortdata[nsortflds] = line;
+	sortdata[nsortflds + 1] = fieldspace;
+
+	free(tmpfields);
+	return sortdata;
+}
+
+char *line(char *sortdata[])
```

Comment:
Is this used? Consider adding `static` to functions you only intend to use in this C file so the compiler can tell you it's unused.

Reply:
It is used, (line 58) to get the original line back from the sortdata after the sort.   

Ah! I think I did try making some functions static but got the error about them being undefined - didn't read the error properly so removed the `static` again :-(

As a _general_ principle I'm guessing if you can make file-level functions and values static you should?  (Not in functions and blocks).